### PR TITLE
WIP: Transform Schur form into eigenvectors

### PR DESCRIPTION
**Pull request by NymanLauri**

### CODE REVIEW

**Comment by haampie** [comment]
Thanks, I'll review

File: src/run.jl [hunk 1]
```diff
@@ -16,14 +16,36 @@ function restarted_arnoldi(A::AbstractMatrix{T}, min = 5, max = 30, nev = min, 
...
+                R[j,j] -= H_copy[i,i]
+            end
+            y[i] = one(T)
+            y[1:i-1] = -R[1:i-1,i]
+            y[i+1:n] = zero(T)
+            # display(y)
+            backward_subst!(view(R,1:i,1:i), y)
+            y ./= norm(y)
+
+            res[i] = abs(dot(view(Q, n, :), y) * arnoldi.H[active + i, active + i - 1]) # H... is this correct?
```
Comment (haampie) [inline src/run.jl]: This can't be correct. Should be `H[max + 1, max]`.

File: src/run.jl [hunk 2]
```diff
@@ -16,14 +16,36 @@ function restarted_arnoldi(A::AbstractMatrix{T}, min = 5, max = 30, nev = min, 
...
+
+        y = Vector{T}(undef,n)
+        res = Vector{Float64}(undef,n)
+        @inbounds for i = n : -1 : 1
+            R = copy(H_copy)
+            @inbounds for j = 1:n
+                R[j,j] -= H_copy[i,i]
+            end
+            y[i] = one(T)
+            y[1:i-1] = -R[1:i-1,i]
```
Comment (haampie) [inline src/run.jl]: Use broadcasting w/ view

File: src/run.jl [hunk 3]
```diff
@@ -16,14 +16,36 @@ function restarted_arnoldi(A::AbstractMatrix{T}, min = 5, max = 30, nev = min, 
...
         H_copy = copy(view(arnoldi.H, active:max, active:max))
         local_schurfact!(H_copy, Q)
-        λs = sort!(eigvalues(H_copy), by = abs, rev = true)
+        # λs = sort!(eigvalues(H_copy), by = abs, rev = true)
+        λs = eigvalues(H_copy)
+
+        y = Vector{T}(undef,n)
+        res = Vector{Float64}(undef,n)
+        @inbounds for i = n : -1 : 1
+            R = copy(H_copy)
```
Comment (haampie) [inline src/run.jl]: No need to make a copy, just work with `H - lambda * I` implicitly; pass H and lambda to the backward substitution function.

**Comment by haampie** [comment]
Please add tests first before you start refactoring! I don't see a test for the backward substitution bit.

File: src/implicit_restart.jl [hunk 4]
```diff
@@ -88,6 +88,10 @@ function single_shift!(H_whole::AbstractMatrix{Tv}, min, max, μ::Tv, Q::Abstrac
     @inbounds H[n, n - 1] = H[n + 1, n - 1]
     @inbounds H[n + 1, n - 1] = zero(Tv)
 
+    # Update Q with the last rotation
```
Comment (haampie) [inline src/implicit_restart.jl]: I don't think this belongs here. We're not using this column?

File: src/implicit_restart.jl [hunk 5]
```diff
@@ -145,6 +149,10 @@ function double_shift!(H_whole::AbstractMatrix{Tv}, min, max, μ::Complex, Q::Ab
         # Zero out the off-diagonal guys
         H[n    , n - 2] = zero(Tv)
         H[n + 1, n - 2] = zero(Tv)
+
```
Comment (haampie) [inline src/implicit_restart.jl]: Same here

File: src/run.jl [hunk 6]
```diff
@@ -16,14 +16,36 @@ function restarted_arnoldi(A::AbstractMatrix{T}, min = 5, max = 30, nev = min, 
...
     for restarts = 1 : max_restarts
+        n = max - active + 1
 
         iterate_arnoldi!(A, arnoldi, min′ + 1 : max, h)
         
         # Compute the eigenvalues of the active part
-        Q = Matrix{T}(I, max, max)
+        Q = Matrix{T}(I, n, n)
         H_copy = copy(view(arnoldi.H, active:max, active:max))
         local_schurfact!(H_copy, Q)
```
Comment (haampie) [inline src/run.jl]: Can you extract this to a function so we can unit test it?

File: src/eigvals.jl [hunk 7]
```diff
@@ -352,12 +352,12 @@ function backward_subst!(R::AbstractMatrix, y::AbstractVector)
...
     n = size(R,1)
 
-    for i = n-1 : -1 : 1
-        R[i,1:i] ./= R[i,i]
-        y[i] ./= R[i,i]
-        for k = i-1 : -1 : 1
-            R[k,1:i] .-= R[i,1:i].*R[k,i]
-            y[k] .-= y[i]*R[k,i]
+    @inbounds for i = n-1 : -1 : 1
+        R[i,1:i] /= R[i,i]
```
Comment (haampie) [inline src/eigvals.jl]: This is not updating R, it's updating a copy of R. Please show things are correct before worrying about performance.

**Comment by haampie** [comment]
Can you also remove the lapack stuff from the PR?

File: src/implicit_restart.jl (hunk 4 above)
Comment (NymanLauri) [reply src/implicit_restart.jl]: Right. Maybe the tests for which these lines are necessary should be removed

Comment (haampie) [reply src/implicit_restart.jl]: You can move it to the tests ;)

**Comment by haampie** [comment]
Starting to look good, test failure is unrelated. Do you have an idea how to do things in real arithmetic here?

**Comment by haampie** [comment]
Btw, you're not including the new test in runtests.jl, so it doesn't run on CI

File: src/run.jl [hunk 8]
```diff
@@ -72,28 +72,80 @@ function transform_converged(arnoldi, active, new_active, min′, V_prealloc)
...
     n = max - active + 1
     # Compute the eigenvalues of the active part
     Q = Matrix{T}(I, n, n)
-    H_copy = copy(view(H, active:max, active:max))
-    local_schurfact!(H_copy, Q)
-    # λs = sort!(eigvalues(H_copy), by = abs, rev = true)
-    λs = eigvalues(H_copy)
+    # vals,vecs = eigen(H[active:max,active:max])
+
+    R = copy(view(H, active:max, active:max))
```
Comment (haampie) [inline src/run.jl]: Just write H[active:max, active:max] here

File: src/run.jl [hunk 9]
```diff
@@ -105,13 +105,13 @@ function compute_shifts(H::AbstractMatrix{T}, active, max) where {T}
...
 
-function compute_shifts(H::AbstractMatrix{T}, active, max) where {T<:Real}
+function compute_shifts(H::AbstractMatrix{T}, active, max, tol=1e-10) where {T<:Real}
     n = max - active + 1
     # Compute the eigenvalues of the active part
     Q = Matrix{T}(I, n, n)
     # vals,vecs = eigen(H[active:max,active:max])
 
-    R = copy(view(H, active:max, active:max))
+    R = copy(H[active:max, active:max])
```
Comment (haampie) [inline src/run.jl]: Remove the copy as well, cause `H[range, range]` is making a copy ^^

File: src/targets.jl [hunk 10]
```diff
@@ -0,0 +1,31 @@
+abstract type Target end
+
+# For finding eigenvalues with the smallest magnitude
+struct SM{T} <: Target end
```
Comment (haampie) [inline src/targets.jl]: Parametrization doesn't make sense

File: src/targets.jl [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+# For finding eigenvalues with the smallest real part
+struct SR{T} <: Target end
+
+# For finding eigenvalues with the largest imaginary part
+struct LI{T} <: Target end
+
+# For finding eigenvalues with the smallest imaginary part
+struct SI{T} <: Target end
+
+sort_vals!(λs,::Type{LM}) = sort!(λs, by=abs, rev=true)
```
Comment (haampie) [inline src/targets.jl]: Just use instances so you don't have to dispatch on `::Type{LM}` but `::LM`

File: src/targets.jl [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+sort_vals!(λs,::Type{SM}) = sort!(λs, by=abs)
+
+sort_vals!(λs,::Type{LR}) = sort!(λs, lt = f(a,b) = abs(real(a)) > abs(real(b)))
```
Comment (haampie) [inline src/targets.jl]: Doesn't make sense to use `abs` here. Also, use `by` not `lt`.

File: src/targets.jl [hunk 13]
```diff
@@ -1,31 +1,31 @@
...
+sort_vals!(λs,::LM) = sort!(λs, by=abs, rev=true)
 
-sort_vals!(λs,::Type{SM}) = sort!(λs, by=abs)
+sort_vals!(λs,::SM) = sort!(λs, by=abs)
 
-sort_vals!(λs,::Type{LR}) = sort!(λs, lt = f(a,b) = abs(real(a)) > abs(real(b)))
+sort_vals!(λs,::LR) = sort!(λs, by = real)
 
-sort_vals!(λs,::Type{SR}) = sort!(λs, lt = f(a,b) = abs(real(a)) < abs(real(b)))
+sort_vals!(λs,::SR) = sort!(λs, by = f(x) = - real(x))
```
Comment (haampie) [inline src/targets.jl]: Either use closures `x -> -real(x)` or use `by = real, rev = true`; doesn't make sense to give this function a name.

**Comment by haampie** [comment]
Needs rebase

**Pull request merged**